Converting Looker PDTs to dbt (part one - reasons why) (#1)
dbanalyticsco/dbanalytics.co: content/posts/looker-pdt-to-dbt-part-one.md

---
title: "Converting Looker PDTs to dbt models: Reasons why (and why not)"
date: "2018-12-25"
author: "Dylan Baker"
---

> Over the past few years, Looker has become a market-leading BI tool. Its data governance and self-service capabilities have now become paramount to thousands of BI stacks. For many, the Third Wave of BI (TM Frank Bien) is well and truly here.

Reviewer: I assume you're going to change the `(TM Frank Bien)` part?

> Over the past few years, Looker has become a market-leading BI tool. Its data governance and self-service capabilities have now become paramount to thousands of BI stacks. For many, the Third Wave of BI (TM Frank Bien) is well and truly here.

Reviewer: I assume you're going to change the `(TM Frank Bien)` part?

Author: I wasn't... But I guess what you're saying is I should?

Author: No. It's not. He just says it in every talk he gives. It was meant as a joke. Clearly hasn't landed though.

> Persistent derived tables (PDTs) are a big part of Looker's success. They allow businesses to quickly transform their data with simple SQL `select` statement, abstracting lots of the data engineering that might otherwise be required. They allow you to easily materialize those models as tables in your data warehouse, speeding up the query times for your end users. 

Reviewer: Fix typo

```suggestion
 Persistent derived tables (PDTs) are a big part of Looker's success. They allow businesses to quickly transform their data with simple SQL `select` statements, abstracting lots of the data engineering that might otherwise be required. They allow you to easily materialize those models as tables in your data warehouse, speeding up the query times for your end users. 
```

> They are addictive. Build a few PDTs, define an explore. Push. Build a few more. Push. Re-define the calculation behind a dimension. Push. Like most addictions though, there comes a point where you realise it may be detrimental to your health and well-being. Well, that of your BI stack anyway. 

Reviewer: For the first "Push." I'd write "Push to production" for clarity. For the subsequent ones, I think "Push." has better rhythm.
```suggestion
 They are addictive. Build a few PDTs, define an explore. Push to production. Build a few more. Push. Re-define the calculation behind a dimension. Push. Like most addictions though, there comes a point where you realise it may be detrimental to your health and well-being. Well, that of your BI stack anyway. 
```

> Models can start getting slow, without the ability to optimise fully. Models can get repetitive, with very portions of code written out over and over. Models can break, without the ability to test their output.

Reviewer: > with very portions of code

^ This doesn't make sense (I assume it's a typo of some sort)

Author: Yeah. What did I mean...?

## Enter: dbt.

> dbt is to PDTs what Reeses is to peanut butter slathered on store-brand chocolate: broadly the same thing, but one far better executed that the other.

Reviewer: Still not loving this analogy.

Maybe... "dbt is to PDTs what good sourdough with cultured butter is to white sandwich bread with margarine" (I think about bread a lot)

Author: I'm going with that.

> From Tristan Handy, the CEO and founder of Fishtown Analytics who build dbt:

Reviewer: I don't love the flow of this sentence. My brain _wants_ to read "build" as "built", because I'm thinking you're taking about Tristan in particular, since he is the subject of the sentence. It's technically correct though, but just a little hard for me to read.

Author: Tristan Handy, founder of dbt-building Fishtown Analytics?

> > dbt is the T in ELT. It doesn’t extract or load data, but it’s extremely good at transforming data that’s already loaded into your warehouse. This “transform after load” architecture is becoming known as ELT (extract, load, transform).

Reviewer: I know this is Tristan's quote, but it's a bit weird. It talks about ELT in the first sentence, and then introduces it as a concept later in the paragraph...

Author: Maybe move the definition to the first instance of ELT?

Author: Or forego/rewrite completely?

Author: Have re-written.

> > Every model is exactly one SELECT query, and this query defines the resulting data set.

Reviewer: Can we put in a reference for this quote?

Author: Yeah.
For my notes: https://blog.fishtownanalytics.com/what-exactly-is-dbt-47ba57309068

It excels at data governance. It excels at self-service. It excels at integrating with other tools. It's just not the best tool for certain types of data modelling.

It may also be right for your business's modelling _right now_. It's a great way to get started with data modelling if you've never done it before. It removes a few steps of complication that having another tool in the mix adds.

So, with that said, here are the instances where I would recommend a move to dbt and the instances where Looker is going to better serve your needs.

## Performance problems and incremental models

Incredibly often, the issue that brings people from PDTs to dbt is the performance of their models. Most of the time, this stems from one of two problems.

The **first** is in inability to [build PDTs incrementally](https://discourse.looker.com/t/incremental-pdts/7620). 

> Imagine you've recently set up Heap, Segment, Snowplow or some other solution for event tracking. Initially, you've only got a few days of data, then a few weeks, then a few months. Every time your PDT needs to rebuild, it rebuilds the _whole_ data set. You're data from day 1 hasn't changed, but it has to be included in the data that gets refreshed. Gradually, this causes the model to rebuild more and more slowly, frustrating your end users who just want to know how many times people viewed the product page for your new special flying mattress yesterday.

Reviewer: Edits: 
* `You're` to `your`
* `1` to `one`

```suggestion
 Imagine you've recently set up Heap, Segment, Snowplow or some other solution for event tracking. Initially, you've only got a few days of data, then a few weeks, then a few months. Every time your PDT needs to rebuild, it rebuilds the _whole_ data set. Your data from day one hasn't changed, but it has to be included in the data that gets refreshed. Gradually, this causes the model to rebuild more and more slowly, frustrating your end users who just want to know how many times people viewed the product page for your new special flying mattress yesterday.
```

Reviewer: Also love the "flying mattress" reference

With dbt, you can build models incrementally, only building or rebuilding the rows that have been created or changed since you last ran the model. This can substantially speed up the build of your models. It will also greatly reduce the load on your database when the models build, allowing other queries to be served more quickly. 

The **second** problem is a tendency to never build intermediary PDTs to speed up model builds.

> There is a tendency in Looker to only build views (and therefore PDTs) that are necessary as a component of an explore. Often though, it would be beneficial for the speed of your rebuilds to break one long model into a handful of tables and then reference those in a final query. `base_pdt` -> `intermediary_pdt` -> `final_pdt` can often be quicker than if all the logic exists in one final larger table. It's not that this can't be done in Looker - it's just that it feel wrong to generate all those views in Looker if they aren't directly for analytics purposes	. 

Reviewer: Edits:
* `than if all the logic exists in one final larger table` -> `than having all the logic in one final larger query`
* `-` -> `—`
* `aren't directly for analytics purposes	.` -> `aren't being surfaced to your end users.`

```suggestion
 There is a tendency in Looker to only build views (and therefore PDTs) that are necessary as a component of an explore. Often though, it would be beneficial for the speed of your rebuilds to break one long model into a handful of tables and then reference those in a final query. `base_pdt` -> `intermediary_pdt` -> `final_pdt` can often be quicker than having all the logic in one final larger query. It's not that this can't be done in Looker — it's just that it feels wrong to generate all those views in Looker if they aren't being surfaced to your end users. 
```

The workflow described above is common place in dbt. Ultimately, it creates a separation between the preparation of data and the LookML that instructs Looker how to expose your final modelled output. 

## Testing your models

Another feature that dbt brings to the table is model testing, similar to unit tests present in most software development languages. 

Looker's data governance capabilities are fantastic. They ensure that everyone is looking at the same data - the same metrics. However, it doesn't ensure that the data feeding those metrics is correct. If someone makes a mistake in the SQL, everyone is still looking at the same metric, that metric just isn't accurate anymore. 

Imagine the following situation. You use Salesforce as your CRM and you have a LookML view called `sf_account`:

{{< highlight sql >}}

select 
	a.id,
	a.name
from salesforce.sf_account a

{{< /highlight >}}

Everything is working correctly. You're completely confident in the data generated in your reports.

> Then, you decide you want to add the stage name from the related Salesforce opportunities to the model. Absent-mindedly, you forget that Accounts -> Opportunities is a one-to-many relationship. You make the following change and all your data from the `sf_account` view is suddently fanned out, distorting metrics in all your reports. 

Reviewer: `suddently` -> `suddenly`
```suggestion
 Then, you decide you want to add the stage name from the related Salesforce opportunities to the model. Absent-mindedly, you forget that Accounts -> Opportunities is a one-to-many relationship. You make the following change and all your data from the `sf_account` view is suddenly fanned out, distorting metrics in all your reports. 
```

{{< highlight sql >}}

select 
	a.id,
	a.name
from salesforce.sf_account a
left join salesforce.sf_opportunity o
	on o.accountid = a.id

{{< /highlight >}}

> The ID column was supposed to be your unique primary key and now it isn't. There is currently no automated mechanism in Looker to test the output of that query. No way of knowing that a record for ID `176` is now there 5 times, one for each opportunity generated for that account.

Reviewer: Edits:
* `knowing that a record for ID `176` is now there 5 times` -> `knowing that ID `176` now has five records`

```suggestion
 The ID column was supposed to be your unique primary key and now it isn't. There is currently no automated mechanism in Looker to test the output of that query. No way of knowing that ID `176` now has five records, one for each opportunity generated for that account.
```

> In dbt, you can define constraints on the model, giving you the ability to identify instances like these in your development workflow.

Reviewer: I really like thinking about tests as `assumptions` rather than constraints. You assert your assumptions about the output, and a test fails when that assumption isn't true.

{{< highlight yaml >}}

- name: sf_account
  columns:
    - name: id
      tests:
        - not_null
        - unique

{{< /highlight >}}

> When you run `dbt test`, you'd realise your mistake and fix it before your Head of Sales starts shouting at you because all the numbers changed. Though testing of that sort hasn't always been common-place in analytics, it's incredibly important so that everyone can have confidence in the reporting they consume.

Reviewer: ```suggestion
 When you run `dbt test`, you'd realise your mistake and fix it before your Head of Sales starts shouting at you because all the numbers changed. Though testing of that sort hasn't always been commonplace in analytics, it's incredibly important so that everyone can have confidence in the reporting they consume.
```

> ## PDT table names

Reviewer: I don't know if this heading is the best way of describing the problem.
Maybe more like "Exposing PDTs to other data users". That doesn't feel right either though.

Author: Yeah I agree. I can't think of something better.

Author: Have tried 'Making data usage agnostic' for the moment. Thoughts?

> David from your data science team wants to build a fancy model, predicting how many flying mattresses you'll sell next week. To do so, he wants to use the `customer_status` dimension defined in your `customers` view in Looker. He doesn't want to have re-define it in his work. He understandably thinks you should all be using the same definition. 

Reviewer: Very tiny thing, but can you change this to a woman's name since DS is a male-dominated industry? It's a pattern I've noticed recently in other blogs and really appreciate.

Reviewer: ```suggestion
 David from your data science team wants to build a fancy model, predicting how many flying mattresses you'll sell next week. To do so, he wants to use the `customer_status` dimension defined in your `customers` view in Looker. He doesn't want to have redefine it in his work. He understandably thinks you should all be using the same definition. 
```

He jumps into his database explorer to find the name of the PDT in the `looker_scratch` schema, the location all the PDTs are built. He finds it. The table is called `lr$zzzxu3bfa1lkc2x5ne75_customers`. He goes off and builds his model. It has an AUC score of 0.96. He's incredibly pleased. He shares the analysis. 

> Your COO opens it up, tries to run it. To his (and David's) dismay, the following error occurs: `ERROR: relation "looker_scratch.lr$zzzxu3bfa1lkc2x5ne75_customers" does not exist`. The table doesn't exist anymore. 

Reviewer: Again here. Since there's two characters, it's OK to go one man, one woman :)

> Every time you make a change to the SQL in a PDT, Looker will rename the underlying table in your database. As is the case here, Looker is often not the only platform or location where data analysis will be produced in your business. Often you'll have a data science function. Often you'll have people who need to write ad-hoc sql queries for specific pieces of analysis. As David initially asserted, all those functions should be building off the same models and the same definitions, building off the same single source of truth. 

Reviewer: > Every time you make a change to the SQL in a PDT, Looker will rename the underlying table in your database.

Is that true? I thought it was every time it refreshed, but I may be wrong here!

Author: Yeah pretty sure it's when the SQL query is changed.

The way Looker builds its PDTs isn't conducive to that principal. The table names will continue to change (and aren't memorable for analysts to use in ad-hoc querying). 

> dbt will build your `customers` model in a table called `customers`. That will never change unless you tell it to. It can be consumed by every user and every platform in perpetuity. David can use it for his analysis. He could even put that model into production because he can be confident that your analytics team controls if that table exists or doesn't.

Reviewer: I feel there needs to be a mention of a schema here, not sure if using `target` is too dbt-jargon-y
```suggestion
 dbt will build your `customers` model in a table called `customers` in your target schema. That will never change unless you tell it to. It can be consumed by every user and every platform in perpetuity. David can use it for his analysis. He could even put that model into production because he can be confident that your analytics team controls if that table exists or doesn't.
```

Author: Given both push to their own schemas, I kinda assumed there is parity of understanding and it didn't need to be mentioned?

Author: Actually, had added and amended slightly.

> 

Reviewer: I think some people actually hack around this problem using the Looker API, which is a significantly more convoluted way of accessing data. I'm not sure if that's worth mentioning or not.

Author: Added a bit.

> ## Moving away from Looker (god forbid)

Reviewer: god? God? Not sure here...

Author: Removed.

That leads me to a slightly meta point. 

Looker is great. I love Looker. I wholly endorse Looker. I don't see that changing any time soon. 

That said, eventually, you or I may choose to move to a different BI platform. It may no longer be the tool best suited for what you are trying to achieve.

At that point, given that the underlying data models need to be used by other users and platforms, you're going to need to retain those models somewhere. Might I suggest dbt?

> ![dbt business intelligence stack!](/dbt-stack-diagram.png)

Reviewer: We have a new version of this diagram! There's a white background one [here](https://raw.githubusercontent.com/fishtown-analytics/dbt/dev/stephen-girard/etc/dbt-arch.png). There's a dark background one as well, but I would have to ask Drew to send it to me.

Author: Changed.

## DRY code

SQL is a notoriously WET (write every time) language. It's veritably soaking with repetition. 

Need to pivot out a column into 5 columns? 5 lines of SQL.

Need to group by 76 columns? (Not sure why.) You need to write out every number between 1 and 76. 

Need to calculate the time between two dates, subtracting for weekends in between? The same lovely SQL snippet over and over and over.

> Most people's SQL is rife with repetition. Most people's SQL is rife with repetition. 

Reviewer: Lol. Was about to correct this.

Reviewer: Lol. Was about to correct this.

Author: hahahaha.

## A quick note

Before I go any further, I want to remark that I wholeheartedly believe that Looker is a fantastic BI tool. It has been transformative for my work at a number of different businesses. It has powered most of the reporting I have done in the past three years. It is great. I fully intend to work with Looker for years to come. I wholly endorse it. 

It excels at data governance. It excels at self-service. It excels at integrating with other tools. It's just not the best tool for certain types of data modelling.

It may also be right for your business's modelling _right now_. It's a great way to get started with data modelling if you've never done it before. It removes a few steps of complication that having another tool in the mix adds.

So, with that said, here are the instances where I would recommend a move to dbt and the instances where Looker is going to better serve your needs.

## Performance problems and incremental models

Incredibly often, the issue that brings people from PDTs to dbt is the performance of their models. Most of the time, this stems from one of two problems.

> The **first** is in inability to [build PDTs incrementally](https://discourse.looker.com/t/incremental-pdts/7620). 

Reviewer: ```suggestion
 The **first** is an inability to [build PDTs incrementally](https://discourse.looker.com/t/incremental-pdts/7620). 
```

> Imagine you've recently set up Heap, Segment, Snowplow or some other solution for event tracking. Initially, you've only got a few days of data, then a few weeks, then a few months. Every time your PDT needs to rebuild, it rebuilds the _whole_ data set. Your data from day one hasn't changed, but it has to be included in the data that gets refreshed. Gradually, this causes the model to rebuild more and more slowly, frustrating your end users who just want to know how many times people viewed the product page for your new special flying mattress yesterday.

Reviewer: ```suggestion
 Imagine you've recently set up Heap, Segment, Snowplow or some other solution for event tracking. You build a PDT that turns these events into page views. Initially, you've only got a few days of data, then a few weeks, then a few months. Every time your PDT needs to rebuild, it rebuilds the _whole_ data set. Your data from day one hasn't changed, but it has to be included in the data that gets refreshed. Gradually, this causes the model to rebuild more and more slowly, frustrating your end users who just want to know how many times people viewed the product page for your new special flying mattress yesterday.
```

With dbt, you can build models incrementally, only building or rebuilding the rows that have been created or changed since you last ran the model. This can substantially speed up the build of your models. It will also greatly reduce the load on your database when the models build, allowing other queries to be served more quickly. 

The **second** problem is a tendency to never build intermediary PDTs to speed up model builds.

> There is a tendency in Looker to only build views (and therefore PDTs) that are necessary as a component of an explore. Often though, it would be beneficial for the speed of your rebuilds to break one long model into a handful of tables and then reference those in a final query. `base_pdt` -> `intermediary_pdt` -> `final_pdt` can often be quicker than having all the logic in one final larger query. It's not that this can't be done in Looker - it's just that it feel wrong to generate all those views in Looker if they aren't being surfaced to your end users. 

Reviewer: ```suggestion
 There is a tendency in Looker to only build views (and therefore PDTs) that are necessary as a component of an explore. Often though, it would be beneficial for the speed of your rebuilds to break one long model into a handful of tables and then reference those in a final query. `base_pdt` -> `intermediary_pdt` -> `final_pdt` can often be quicker than having all the logic in one final larger query. It's not that this can't be done in Looker - it's just that it feels wrong to generate all those views in Looker if they aren't being surfaced to your end users. 
```

{{< highlight sql >}}

select 
	a.id,
	a.name
from salesforce.sf_account a
left join salesforce.sf_opportunity o
	on o.accountid = a.id

{{< /highlight >}}

The ID column was supposed to be your unique primary key and now it isn't. There is currently no automated mechanism in Looker to test the output of that query. No way of knowing that a record for ID `176` now has five records, one for each opportunity generated for that account.

In dbt, you can define assumptions on the model, giving you the ability to identify instances like these in your development workflow.

{{< highlight yaml >}}

- name: sf_account
  columns:
    - name: id
      tests:
        - not_null
        - unique

{{< /highlight >}}

When you run `dbt test`, you'd realise your mistake and fix it before your Head of Sales starts shouting at you because all the numbers changed. Though testing of that sort hasn't always been commonplace in analytics, it's incredibly important so that everyone can have confidence in the reporting they consume.

## Making data usage-agnostic

> Claire from your data science team wants to build a fancy model, predicting how many flying mattresses you'll sell next week. To do so, she wants to use the `customer_status` dimension defined in your `customers` view in Looker. She doesn't want to have re-define it in her work. She understandably thinks you should all be using the same definition. 

Reviewer: Hahaha! Oh nooooo! 🙈

The way Looker builds its PDTs isn't conducive to that principal. The table names will continue to change (and aren't memorable for analysts to use in ad-hoc querying).

dbt will build your `customers` model in a table called `customers` in the schema of your choice (`target` in dbt parlance). That will never change unless you tell it to. It can be consumed by every user and every platform in perpetuity. Claire can use it for her analysis. She could even put that model into production because she can be confident that your analytics team controls if that table exists or doesn't.

One way around this with Looker is having people build models on top of the Looker API. This can work well, but still requires the data being exposed together in an explore. With external models, people can join up datasets that aren't otherwise modelled together in Looker.

> ## Moving away from looker

Reviewer: ```suggestion
 ## Moving away from Looker
```
